+ Requirements:
    1. The estimated time for performing this activity is approximately half an hour.
<br><br>
+ Objectives of this notebook:
    1. Introduction to OpenCV: learning to acquire images.
<br><br>
+ Summary of activities:
    1. The histogram of an image will be analyzed using Python code snippets for acquired images.

%%latex
\tableofcontents
\newpage

# Introduction to OpenCV
In Python you always start by importing all the required modules:

In [1]:
import numpy as np  # the numpy module is associated with the alias np
import cv2          # OpenCV module
import matplotlib.pyplot as plt  # the matplotlib.pyplot with the alias plt
import time                      # module to mesuare time cost

In [2]:
# **** It is important to consult the help of the version you are using ****
!python --version
print('Numpy:',np.__version__)
print('OpenCv:',cv2.__version__)

Python 3.9.1
Numpy: 1.20.2
OpenCv: 4.5.2


## Select video device

In [3]:
#---------------------------------------------------------------------------
# If you have only one camera, try capturing from device 0
# Each webcam is assigned a numeric identifier
# If you have more cameras, try other devices 1, 2, ...
#---------------------------------------------------------------------------
#camara_dev = 0

#---------------------------------------------------------------------------
# If you have a camera on your cell phone, install an IP camera server.
# In the case of android, install the application "smart webcam".
#
# https://www.portalprogramas.com/smart-webcam/android/
#
# When you run this application on the cell phone, it creates an IP camera handler
# with a specific ip address and port.
# In my case, "private ip": 192.168.1.200 (0) and "Port": 8082
# Modify the following device in accordance with your server
#---------------------------------------------------------------------------
#camara_dev = 'http://10.5.196.92:8082'
camara_dev = 'http://VxC:vxc2018@158.42.206.2/mjpeg/snap.cgi?chn=0'
    
print('Selected device:', camara_dev)

Selected device: http://VxC:vxc2018@158.42.206.2/mjpeg/snap.cgi?chn=0


## Image acquisition

In [4]:
# Open device
video = cv2.VideoCapture(camara_dev)

#video.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)
#video.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)

#video.set(cv2.CAP_PROP_EXPOSURE, -5.0) # 2^-5s = 1/32s
#video.set(cv2.CAP_PROP_EXPOSURE, -8.0)

# Image acquisition: a BGR image is captured
ret_ok, img = video.read()

# Close device
video.release()

if ret_ok:
    print('Image shape: ', img.shape)
    print('Image type:',type(img),'of', img.dtype)
else:
    print('Error: the selected device is not available.')

Image shape:  (720, 1280, 3)
Image type: <class 'numpy.ndarray'> of uint8


## Shows image

In [5]:
cv2.imshow('Image', img) # Opens a window to show the image
cv2.waitKey(0)           # Waits until key pressed on the window Image
cv2.destroyAllWindows()  # Closes the window

## Shows image continously

In [ ]:
video = cv2.VideoCapture(camara_dev)

cv2.namedWindow('Image') # Opens a window to show the image

key = -1 # to start key must be different to ord('q') = 113

while(video.isOpened() and (key != ord('q'))):

    ret_ok, img = video.read()

    if ret_ok:
        cv2.imshow('Image', img)
        key = cv2.waitKey(1)     # press 'q' key to fininsh the while loop

    else:
        print('End of video.')
        break

video.release()
cv2.destroyAllWindows()

## Shows RGB histograms

In [ ]:
# Split the BGR image into three planes
blue, green, red = cv2.split(img)

# Shows RGB histograms
hist_r = np.zeros((256,))
hist_g = np.zeros((256,))
hist_b = np.zeros((256,))

height, width, channels = img.shape
for x in range(height):
    for y in range(width):
        hist_b[img[x,y,0]] = hist_b[img[x,y,0]] + 1
        hist_g[img[x,y,1]] = hist_g[img[x,y,1]] + 1
        hist_r[img[x,y,2]] = hist_r[img[x,y,2]] + 1

plt.plot(hist_r, color='red')
plt.plot(hist_g, color='green')
plt.plot(hist_b, color = 'blue')
plt.show()

# References
+ [Python documentation](https://docs.python.org/3/)
+ [Numpy documentation](https://numpy.org/doc/)
+ [Documentación de OpenCV 4.5](https://docs.opencv.org/4.5.0/)
+ [Machine Learning for OpenCV 4 - Second Edition
by Vishwesh Ravi Shrimali; Michael Beyeler; Aditya Sharma
Published by Packt Publishing, 2019](https://learning.oreilly.com/library/view/machine-learning-for/9781789536300/cover.xhtml)
+ [Machine Learning for OpenCV 4 : Intelligent Algorithms for Building Image Processing Apps Using OpenCV 4, Python, and Scikit-Learn. by Sharma, Aditya, et al. Second edition., 201](https://polibuscador.upv.es/permalink/f/6ptg6i/alma5191663190003706)